### PR TITLE
[SR-5918] Add diagnostic error for duplicate product names

### PR DESCRIPTION
**Pull request by BalestraPatrick**
Fixes [SR-5918](https://bugs.swift.org/browse/SR-5918) by @lightsprint09, @hartbit and myself at [Swift Alps](https://theswiftalps.com).

### CODE REVIEW

**Comment by ddunbar** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/PackageGraphLoader.swift [hunk 1]
```diff
@@ -211,6 +211,12 @@ private func createResolvedPackages(
             .filter({ $0.product.type != .test })
         let productDependencyMap = productDependencies.createDictionary({ ($0.product.name, $0) })
 
+        // Establish if there are multiple products with the same name.
+        let duplicateProductNames = Set(packageBuilder.products.map { $0.product.name }.findDuplicates())
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/PackageGraphLoader.swift]: Do we really require `Set()` here? I don't think findDuplicates should return duplicate elements.

File: Tests/PackageGraphTests/PackageGraphTests.swift [hunk 2]
```diff
@@ -136,6 +136,35 @@ class PackageGraphTests: XCTestCase {
...
+                    .library(name: "Boo", targets: ["Bar"]),
+                    .library(name: "Boo", targets: ["Bar"])
+                ],
+                targets: [
+                    .target(name: "Bar"),
+                ]),
+            ], root: "/Foo", diagnostics: diagnostics, in: fs)
+
+        let multipleBarDiagnostics = diagnostics.diagnostics.filter { $0.localizedDescription == "multiple products named 'Bar'" }
+        let multipleBooDiagnostics = diagnostics.diagnostics.filter { $0.localizedDescription == "multiple products named 'Boo'" }
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PackageGraphTests.swift]: We should also report the package name in this error.

File: Sources/PackageGraph/PackageGraphLoader.swift (hunk 1 above)
Comment (aciidb0mb3r) [reply Sources/PackageGraph/PackageGraphLoader.swift]: This won't look for the duplicate products across packages, we should probably do that.

File: Tests/PackageGraphTests/PackageGraphTests.swift [hunk 3]
```diff
@@ -136,6 +136,35 @@ class PackageGraphTests: XCTestCase {
...
+
+        XCTAssertEqual(multipleBarDiagnostics.count, 1)
+        XCTAssertEqual(multipleBooDiagnostics.count, 1)
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PackageGraphTests.swift]: This can be improved by adding an utitlity method for checking for diagnostics.

**Comment by DougGregor** [comment]
This is causing breakage in the source compatibility suite. Bug at:

  https://bugs.swift.org/browse/SR-6488

Perhaps we should stage this in as a warning?

**Comment by aciidb0mb3r** [comment]
I think source compatibility != package manager compatibility. Its probably a good thing that package is breaking now. Can we disable that package from the suite until the author submits an update?

**Comment by aciidb0mb3r** [comment]
Technically its not a problem to make it a warning. However, this is something which _may_ cause undefined behaviour. I'd rather force users to fix their package than continue defining it the wrong way. Usually, we try to maintain as much package manager compatibility as possible (we even ship two versions of package manifest API right now) but in certain cases it is better to just break packages. I don't expect this particular issue to break a large number of packages, so I don't think its worth turning this into a warning and then to an error later.

Having said that, policy for breaking package manager changes w.r.t the source compatibility suite is something we need to discuss.

**Pull request merged**